oci/casext/mediatype/parse.go

```diff
@@ -136,22 +137,66 @@ func CustomJSONParser(v interface{}) ParseFunc {
 	if t.Kind() != reflect.Struct {
 		panic("CustomJSONParser() called with non-struct kind!")
 	}
-	return func(reader io.Reader) (_ interface{}, err error) {
+	return func(rdr io.Reader) (_ interface{}, err error) {
 		ptr := reflect.New(t)
-		if reader != nil {
-			err = json.NewDecoder(reader).Decode(ptr.Interface())
+		if rdr != nil {
+			err = json.NewDecoder(rdr).Decode(ptr.Interface())
 		}
 		ret := reflect.Indirect(ptr)
 		return ret.Interface(), err
 	}
 }
 
+func indexParser(rdr io.Reader) (interface{}, error) {
+	// Construct a fake struct which contains bad fields. CVE-2021-41190
+	var index struct {
+		ispec.Index
+		Config json.RawMessage `json:"config,omitempty"`
+		Layers json.RawMessage `json:"layers,omitempty"`
+	}
+	if rdr != nil {
+		if err := json.NewDecoder(rdr).Decode(&index); err != nil {
+			return nil, err
+		}
+	}
+	if index.MediaType != "" && index.MediaType != ispec.MediaTypeImageIndex {
+		return nil, fmt.Errorf("malicious image detected: index contained incorrect mediaType: %s", index.MediaType)
+	}
+	if len(index.Config) != 0 {
+		return nil, fmt.Errorf("malicious image detected: index contained forbidden 'config' field")
```

Comment:
For all these it seems like maybe we want errors.Errorf (and an errors.Wrapf()) so that users get stack traces?